src/updateSlice.js

```diff
@@ -0,0 +1,35 @@
+import { curry } from 'ramda';
+
+import concatAll from './concatAll';
+import sliceFrom from './sliceFrom';
+import sliceTo from './sliceTo';
+
+/**
+ * Returns target with given slice replaced using passed replacement.
+ *
+ * @func updateSlice
+ * @memberOf RA
+ * @category List
+ * @sig Number -> Number -> [a] -> [a] -> [a]
+ * @sig Number -> Number -> String -> String -> String
+ * @param {number} fromIndex
+ * @param {number} toIndex
+ * @param {string|any[]} replacement
+ * @param {string|any[]} target
+ * @returns {string|any[]} target with given slice replaced using passed replacement
+ * @example
+ *
+ * RA.updateSlice(1, 3, 'DDDD', 'ABBCCC') //=> 'ADDDDCCC'
+ * RA.updateSlice(1, 3, Array.from('DDDD'), Array.from('ABBCCC')) //=> [ 'A', 'D', 'D', 'D', 'D', 'C', 'C', 'C' ]
+ * RA.updateSlice(-5, -3, 'DDDD', 'ABBCCC') //=> 'ADDDDCCC'
+ * RA.updateSlice(3, 1, 'DDDD', 'ABBCCC') //=> 'ABBDDDDBBCCC'
```

Comment:
This doesn't make sense to me. See how [R.slice](http://ramdajs.com/docs/#slice) handles this case. IMHO we should return empty array or empty string in these cases.

See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice
and https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice

Reply:
replace empty selection = return empty array? That doesn't seem right.

Python treats it as insertion.
I was about to say that we don't need it, until I noticed that `R.insertAll` works only for arrays.

Default behavior is not very useful, IMO, we'd need to special case it if we want to support it.

Reply:
Upon reflection, I agree that returning empty array or string is confusing and doesn't seem correct. Alternative is to return the original list unmodified or throw. 

@Undistraction, @guillaumearm opinions on this ?

Reply:
@Undistraction, @guillaumearm just pinging you, take a look at this when you have time.

Reply:
I think we should have same behavior as `RA.inRange` and throw.
IMO, having to handle this case is weird.

Reply:
It's a tricky one. Current behaviour is definitely wrong. `R.slice` will return an empty list if both ends of the supplied range don't sit within the list, but here that behaviour doesn't make much sense. I would say return the list untouched, but what if there is a partial overlap? In JS doesn't even support negative indexes so shouldn't we should clamp them to zero anyway? As for the top end, if the upper index is over, we replace the whole thing. If the lower index is over we return the list untouched.

Reply:
@guillaumearm no you understood it correctly. So we have two sensible approaches. Throw or return original string.

Reply:
We need to handle:

A. both negative indices
B. 1. negative index, 2. positive within range
C. 1. positive within range, 2. negative index
D. 1. positive within range, 2. positive out of range
E. both positive out of range.

I think different situations require different results.

- We can coerce negative indices to 0 which will work for A (returns original list) and B (replaces from 0 - second index).
- C should throw as this is completely invalid
- D should replace and extend
- E should return the original list

What do you think?

Reply:
@Undistraction I wouldn't coerce negative indexes to 0. Negative indexes are valid, look at tests, they count from the end. But, it's a valid concern, `fromIndex` may be greater that `toIndex`, because `toIndex` might be negative.

What's worse, wheter there will be overlap or not depends on length of `list`...

Reply:
>  I wouldn't coerce negative indexes to 0. Negative indexes are valid, look at tests, they count from the end

Is that something we want? If we want consistency (with Ramda's clamp), we should error here. This also applies to `inRange` Do we want to support the idea of negative indexes? At present we don't in `inRange`.

Reply:
`inRange` supports negative indexes: `const isSingleDigit inRange(-9, 10)`

Reply:
`inRange` is a different case because we are just talking about a range of numbers and numbers can be negative. So -x literally means -x. In this function (in the current implementation -x doesn't literally mean -x, it means the index that is -x from the last item in the list. 

Personally I find this behaviour pretty confusing, however given that a negative index wouldn't make sense at all when used on a list or string in any other context, I think adding this isn't so much of a problem - we can trust that a user probably hasn't accidentally used a negative index. 

So now that I understand the negatives - that they are just a mirror of positives, counting from the last item, the list becomes what happens if:

A.1 `fromIndex` is negative and abs(fromIndex) is greater than list length
A.2 `fromIndex` is positive and greater than list length
B.1 `toIndex` is negative and abs(toIndex) is greater than list length
B.1 `toIndex` is positive and  greater than list length

However, this is assuming that negative indexes are always counted from the end of the list. The tests do not make clear what should happen if `fromIndex` is negative and `toIndex` is positive and vice-versa. So what is the correct behaviour is when:

1. `fromIndex` is positive and `toIndex` is negative.
2. `fromIndex` is negative and `toIndex` is positive.

Reply:
I'd make a function `absIndex` taking index and array or length, and returning normal index. That allows to just (pseudocode) `if absIndex(from) > absIndex(to) throw`.

Reply:
The `abs` is just shorthand. Forget the implementation. What should the behaviour be?

Reply:
Behavior of what?

Reply:
@GingerPlusPlus behavior of the `updateSlice` util in cases that @Undistraction described.